### ISSUE TITLE
grouped_mm backward NaN on Blackwell (sm_103) in full-scale MoE EP training (torch 2.13.0+cu130)

### ISSUE THREAD

**Issue by pasta-paul** [pytorch/pytorch#193633]
## Summary

In full-scale MoE expert-parallel (EP=8) training of a 219B-param Cohere2Moe-class model on 8x NVIDIA B300 (sm_103) with torch 2.13.0+cu130, the `grouped_mm` experts path produces NaN gradients across essentially all parameters starting at the **second** backward pass. The first backward is clean. The mathematically identical eager (per-expert loop) experts path trains cleanly for hundreds of steps. A "compaction" workaround (dropping the zero-count sentinel tail before the grouped projections) has passed a full-scale production gate: 3+ optimizer steps, losses matching the eager band at the same steps, zero NaN.

## Environment

- torch 2.13.0+cu130 (pip wheel)
- 8x NVIDIA B300 SXM (compute capability 10.3, sm_103), NVLink, single node
- Model: 219.33B-param MoE (Cohere2Moe class), 128 experts, top-8 routing, EP=8 via torch.distributed DeviceMesh, bf16 params, no gradient checkpointing
- Workload: 1,048,576 tokens per optimizer step (micro-batches of 4x2048 tokens, grad accumulation)

## Production symptom

- backward #1: all gradients finite
- backward #2 onward: NaN gradients across ~all parameter categories (experts, router, attention, embeddings)
- Same checkpoint, same data, same hyperparameters with the eager experts path: completely clean
- 100% reproducible at production scale (3/3 grouped_mm attempts NaN'd; 120+ steps eager clean)

## Hypotheses tested and eliminated (all clean)

| # | Hypothesis | Test setup | Result |
|---|---|---|---|
| 1 | Sentinel tail rows (uninitialized memory beyond `offsets[-1]`) corrupt real rows | tail present; tail explicitly NaN-poisoned | clean 4/4 fwd+bwd |
| 2 | Non-contiguous / transposed-B operand | NaN-poisoned grad tail + transposed B | clean 4/4 |
| 3 | Zero-count groups (empty experts, cf. cutlass stale-accumulator family) | 4 empty experts, dirtied caching allocator, 6 iterations | clean; empty experts got exactly-zero grads |
| 4 | Small-scale EP dispatch | 2-rank tiny-EP with real dims (H=4096, I=4096, 16 experts), RouterParallel dispatch, DTensor grads | clean |

So the trigger is NOT any of: sentinel tails, non-contiguous operands, zero-count groups, or small-scale EP dispatch. It requires full scale (or some other property of the production config) to manifest. Repro scripts for all eliminated hypotheses are available on request.

## Workaround (passed full-scale production gate)

Compaction: before each grouped projection, slice the sorted token buffer to the `offsets[-1]` real rows (one host sync per layer), run grouped_mm on the tail-free tensors, scatter results back into a zeros buffer afterward. With this patch:

- 3 gated production steps: losses 1.1959 / 1.2003 / 1.2189 vs eager reference 1.2004 / 1.2188 / 1.1705 at the same steps
- zero NaN/Traceback
- throughput 5,913 tok/s vs 1,894 tok/s eager (3.12x)

## Related

- #154557 (grouped_mm optional zero initialization of the output)
- NVIDIA/cutlass#2674 (zero-M groups stale-accumulator family)
- #172439 (grouped_mm fixes in torch >= 2.10)

Happy to run targeted experiments on the production node if a maintainer has a hypothesis to test -- the setup reproduces the NaN reliably without the compaction patch.

---

*This report was prepared with Digby AI assistance (digby.ai) -- investigation, repro scans, and workaround verification performed autonomously on a production 8xB300 node.*

cc @ezyang @gchanan @kadeng @msaroufim @ptrblck @eqy @tinglvv @nWEIdia @jianyuh @nikitaved @mruberry @walterddr @Lezcano

**Comment by soulitzer** [comment]
cc @eqy

**Comment by drisspg** [comment]
Hey, I did take a look at this

**TL;DR:** I couldn't reproduce any kernel-level miscompute, but I can deterministically reproduce your exact failure signature (backward #1 clean → backward #2 NaN), and I think it's due to consumption of the **uninitialized padding rows beyond `offsets[-1]`**.

### What I think is happening

`_grouped_mm` allocates its output with `empty_strided` and only writes rows `< offs[-1]`. **Note: this is expected behavior and we don't have plans to change it** (see #154557). It applies not just to the forward output, but also to **dX produced inside autograd** (`_grouped_mm_mat1_backward`) which user code can't see or sanitize:

- Backward #1: dX comes from fresh `cudaMalloc` pages, which read as zeros → clean.
- Backward #2+: the caching allocator recycles blocks that previously held freed bf16 activations. The tail is now arbitrary bit patterns, including NaN/inf → any downstream consumer of those rows poisons real gradients. Note `0 × NaN = NaN`, so a zero routing-weight multiply does not protect you — this is likely why your poisoning tests looked clean.

I verified on GB200 (same CUTLASS sm10x dispatch path as your B300) with your exact wheel (2.13.0+cu130): valid regions are always correct across an extensive stress/fuzz sweep, while a deterministic allocator-poisoning repro produces a dX tail that is 100% NaN on the recycled iteration with the valid region still exact.

### Quick confirmation / hotfix

You can add a gradient-sanitizing hook on each padded buffer feeding a grouped_mm —> if this makes the NaNs vanish, the diagnosis is confirmed:

```python
row = torch.arange(x.size(0), device=x.device)
x.register_hook(lambda g: g.masked_fill((row >= offs[-1]).unsqueeze(-1), 0))
```

### The recommended fix

This comes up alot and I have a worked KDA training example ([usage](https://github.com/meta-pytorch/attention-gym/blob/17796cd6b3c6dfc52e5f60a6b6c267a697bee8ce/examples/kda_training.py#L236-L260), [helpers: `active_token_mask` / `mask_inactive_tokens` / `mask_inactive_token_gradients`](https://github.com/meta-pytorch/attention-gym/blob/17796cd6b3c6dfc52e5f60a6b6c267a697bee8ce/attn_gym/linear/kda/masking.py#L89-L150)):

```python
active = (torch.arange(cap, device=dev) < offs[-1]).unsqueeze(-1)  # on-device, no sync

def mask_inactive_tokens(x):            # zero padding VALUES (and their grads)
    return torch.where(active, x, 0)

def mask_inactive_token_gradients(x):   # identity forward, zero padding GRADS
    return torch.where(active, x, x.detach())
```

Placement for a grouped-mm expert block:

```python
x = mask_inactive_token_gradients(x)    # barrier: catches the uninit dX tail
h = torch._grouped_mm(x, w1, offs)
h = mask_inactive_tokens(h)             # uninit fwd tail -> real zeros
h = act(h)
h = mask_inactive_token_gradients(h)
out = torch._grouped_mm(h, w2, offs)
out = mask_inactive_tokens(out)
```
